std/experimental/checkedint.d

```diff
@@ -472,6 +472,23 @@ if (isIntegral!T || is(T == Checked!(U, H), U, H))
         assert(!MyHook.thereWereErrors);
     }
 
+    // toHash
+    /** Generates a hash for `this`. If `Hook` defines `hookToHash`, the call
+      immediately returns `hook.hookToHash(get)`. Otherwise, if payload type
+      defines `toHash`, call is redirected to payload's `toHash`, else the
+      built-in function `getHash` is called.
+    */
+
+    size_t toHash()
+    {
+        static if (hasMember!(Hook, "hookToHash"))
+            return hook.hookToHash(payload);
+        else static if (hasMember!(typeof(payload), "toHash"))
+            return payload.toHash();
+        else
+            return (() @trusted => typeid(payload).getHash(&payload))();
```

Comment:
The problem with doing it this way is `checked!Warn(42).toHash == checked!Abort(42).toHash;`

See https://github.com/dlang/phobos/pull/6457/files for some idea of how to use the hash of the payload + the hash of the Hook

Reply:
I believe this is the right way to do it in order to be consistent with `Checked.opEquals`.
I added an assert to the example provided by @n8sh [here](https://run.dlang.io/is/BmOoYK)

imho, delegating to the hook the behaviour of the has is the most transparent to the user

Reply:
>imho, delegating to the hook the behaviour of the has is the most transparent to the user

If the desired behavior is that `Checked`s with the same payload but different hooks have the same hash code, then it is never correct for a hook to have a custom hash function.

Reply:
Although with implicit numerical promotion it is not possible for `a == b` to entail `hashOf(a) == hashOf(b)`.

https://run.dlang.io/is/gRxZib
```d
void main()
{
    immutable int a = -1;
    immutable long b = -1;
    immutable float c = -1;
    assert(a == b);
    assert(a == c);
    assert(hashOf(a) != hashOf(b));
    assert(hashOf(a) != hashOf(c));
    
    import std.experimental.checkedint : Checked;
    immutable ca = Checked!int(a);
    immutable cb = Checked!long(b);
    assert(ca == cb);
    assert(ca == a && ca == b);
    assert(cb == a && cb == b);
    assert(hashOf(ca) != hashOf(cb) || hashOf(ca) != hashOf(a)
           || hashOf(cb) != hashOf(b), "Cannot all be true.");
    
    //assert(ca == c); // Bonus (doesn't compile).
}
```